### PR TITLE
Thread class cleanup.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
Change to using `std::function` and remove the extra data argument.
For #1997.

### CODE REVIEW

File: iocore/eventsystem/Thread.cc [hunk 1]
```diff
@@ -55,9 +55,8 @@ Thread::~Thread()
 }
 
 static void
-key_destructor(void *value)
+key_destructor(void *)
```
Comment (jpeach) [inline iocore/eventsystem/Thread.cc]: The destructor for `pthread_key_create` is optional, so we can nuke this altogether.

File: iocore/eventsystem/I_Thread.h [hunk 2]
```diff
@@ -63,15 +63,20 @@
...
 class ProxyMutex;
-typedef void *(*ThreadFunction)(void *arg);
 
-static const int MAX_THREAD_NAME_LENGTH = 16;
+constexpr int MAX_THREAD_NAME_LENGTH = 16;
+
+#define THREADAPI
+
+using THREAD_RETURN_TYPE = void *;
+using ThreadFunction     = std::function<THREAD_RETURN_TYPE THREADAPI()>;
```
Comment (jpeach) [inline iocore/eventsystem/I_Thread.h]: The return from the thread function isn't used so we can drop `THREAD_RETURN_TYPE` and just do:
```C
using ThreadFunction = std::function<void()>;
```

File: iocore/eventsystem/Thread.cc [hunk 3]
```diff
@@ -72,42 +71,40 @@ init_thread_key()
...
 static void *
 spawn_thread_internal(void *a)
 {
-  thread_data_internal *p = (thread_data_internal *)a;
+  auto *p = static_cast<thread_data_internal *>(a);
 
   p->me->set_specific();
   ink_set_thread_name(p->name);
+
   if (p->f) {
```
Comment (jpeach) [inline iocore/eventsystem/Thread.cc]: Now that the thread function is `ThreadFunction` value, can this condition ever be false?

File: iocore/eventsystem/I_Thread.h [hunk 4]
```diff
@@ -138,18 +146,10 @@ class Thread
...
   ProxyAllocator ioBlockAllocator;
 
-private:
-  // prevent unauthorized copies (Not implemented)
-  Thread(const Thread &);
-  Thread &operator=(const Thread &);
-
 public:
-  ink_thread start(const char *name, size_t stacksize, ThreadFunction f, void *a, void *stack);
+  ink_thread start(const char *name, void *stack, size_t stacksize, ThreadFunction const &f = ThreadFunction());
```
Comment (jpeach) [inline iocore/eventsystem/I_Thread.h]: I can't find a place where this the `ThreadFunction` is actually used. Unless you are planning to start using it in a subsequent patch, let's remove it?

Comment (SolidWallOfCode) [inline iocore/eventsystem/I_Thread.h]: That's the problem with splitting this up, there are pieces that are needed later but not directly applicable now.

File: iocore/eventsystem/Thread.cc (hunk 3 above)
Comment (SolidWallOfCode) [inline iocore/eventsystem/Thread.cc]: It is false if no function has been stored in it, just like `std::shared_ptr`.

**Comment by jpeach** [comment]
> On Jun 6, 2017, at 6:15 AM, Alan M. Carroll <notifications@github.com> wrote:
> 
> @SolidWallOfCode commented on this pull request.
> 
> In iocore/eventsystem/I_Thread.h:
> 
> >  public:
> -  ink_thread start(const char *name, size_t stacksize, ThreadFunction f, void *a, void *stack);
> +  ink_thread start(const char *name, void *stack, size_t stacksize, ThreadFunction const &f = ThreadFunction());
> 
> That's the problem with splitting this up, there are pieces that are needed later but not directly applicable now.

Ah, that make sense. A note in the commit message is probably the best way to convey this context.

**Pull request merged**